# Final script for using F4-ratio estimation as an orthogonal method to estimate the anubis ancestry proportions of Amboseli individuals (see Supplementary Methods 8)

# Load R libraries
library(ggplot2)
library(data.table)
library(ggrepel)

# Load the R data results for each chromosome generated by the script 06.f4ratio_admixr.R and concatenate into a single data frame called f4
for (i in seq(1,19)) {
  
  if(i<2 | (i>2 & i<10)) {
    load(paste("f4_ratio_chr0", i, ".Rd", sep=""), verbose=T)
  }
  
  if(i==2) {
    load(paste("f4_ratio_chr0", i, "a.Rd", sep=""), verbose=T)
    result_all -> tmp
    load(paste("f4_ratio_chr0", i, "b.Rd", sep=""), verbose=T)
    rbind(result_all, tmp) -> result_all
  }
  
  
  if(i>=10) {
    load(paste("f4_ratio_chr", i, ".Rd", sep=""), verbose=T)
  }
  
  if (i==1) {result_all -> f4}
  if (i>1) {rbind(f4, result_all) -> f4}
  
}

# Make a column for labelling the phylogenetic configurations following the nomenclature of Patterson et al. 2012 Genetics (excluding the focal population X)
# population X is a combination of two sources, populations B' and C'. Because B' and C' are unknown ancestral populations, they are represented by modern populations B and C. This method also requires two additional populations: population A, which has not contributed to population X but is a sister group to population B, and an outgroup, O, to populations A, B, and C. 
f4$phylogeny_noX <- paste(f4$A, f4$B, f4$C, f4$O, sep="/")

# Plot the results (alpha +/- 2 SE) for all phoylogenic configurations colored by population X (Amboseli historically admixed individuals or Amboseli recently admixed individuals)
ggplot() + geom_pointrange(data=subset(f4, X=="Amboseli_recent"), aes(x=phylogeny_noX, y=alpha, ymin = alpha - (2 * stderr), ymax = alpha + (2 * stderr)), color="#009E73") + geom_pointrange(data=subset(f4, X=="Amboseli_historic"), aes(x=phylogeny_noX, y=alpha, ymin = alpha - (2 * stderr), ymax = alpha + (2 * stderr)), color="#FFED4F") +  geom_hline(yintercept = 0, linetype = 2) + facet_wrap(~chrom) + labs(y = "anubis ancestry proportion", x = "phylogeny (A, B, C, Outgroup)") + theme_classic() + scale_x_discrete(labels=c("Guinea/SW_olive/Mikumi/Gelada" = "Guinea/SW_olive/\nMikumi/Gelada",  "Guinea/SW_olive/Mikumi/macaque" = "Guinea/SNPRC_anubis/\nMikumi/macaque", "Hamadryas/SNPRC_anubis/Mikumi/Gelada" = "Hamadryas/SNPRC_anubis/\nMikumi/Gelada", "Hamadryas/SNPRC_anubis/Mikumi/macaque"="Hamadryas/SNPRC_anubis/\nMikumi/macaque")) + geom_text(data=f4, aes(x=phylogeny_noX, y=alpha, label=Zscore), size=3, nudge_x=0.4) + theme(axis.text.x = element_text(size=6.5))

# Get a weighted average based on chromosome size
# Get chromosome base length for Macam genome
macam_length <- read.table("MacaM.fa.fai", header=F)

# Get the columns you need (chromosome = column 1, chromosome length = column 2)
macam_length <- macam_length[c(1:2)]
# Name the columns 
colnames(macam_length) <- c("chrom", "chrom_length")

# Get an average per chromosome per Amboseli hybrid type across the phylogenetic configurations
tmp <- f4 %>% group_by(chrom, X) %>% summarize(mean=mean(alpha), sd=sd(alpha)) 

summary(tmp$mean)
summary(tmp$sd)
round(min(tmp$sd),3) # 0.05
round(max(tmp$sd),3) # 0.037

# Plot the average alpha (anubis ancestry proportion) per chromosome
ggplot() + geom_point(data=subset(tmp, X=="Amboseli_recent"), aes(x=chrom, y=mean), color="#009E73", size=4) + geom_point(data=subset(tmp, X=="Amboseli_historic"), aes(x=chrom, y=mean), color="#FFED4F", size=4) +  geom_hline(yintercept = 0, linetype = 2)  + labs(y = "mean anubis ancestry proportion\nacross phylogenetic configurations", x = "chromosome") + theme_classic() + theme(text=element_text(size=20), axis.text = element_text(color="black"))

# Merge the average alphas across phylogenetic configurations per chromosome with the chromosome length data
tmp2 <- merge(tmp, macam_length, by=c("chrom"))

# Multiple the average alpha across phylogenetic configurations per chromosome by the chromosome length
tmp2$mean_by_chrom_length <- tmp2$mean*tmp2$chrom_length

# Separate Amboseli historically admixed individuals and Amboseli recently admixed individuals into two separate data frames
amb_recent <- tmp2[tmp2$X=="Amboseli_recent",]
amb_historic <- tmp2[tmp2$X=="Amboseli_historic",]

# Sum the average alpha by chromosome length divided by the entire Macam genome length
round(sum(amb_recent$mean_by_chrom_length)/sum(amb_recent$chrom_length), 3) # 0.508
round(sum(amb_historic$mean_by_chrom_length)/sum(amb_historic$chrom_length),3) #  0.232

# Save the final data frames
save(amb_recent, amb_historic, f4, file="f4_results_final.Rd")
